### ISSUE TITLE
Manual entries are not persistant (ipset)

### ISSUE THREAD

**Issue by srulikuk** [fail2ban/fail2ban#2647]
Fail2Ban v0.10.5 - centOS
I am using IPSet with my jail.local's they work fine however I have 1 jail that I use when I manually ban a IP and those IP's are not persistent after restarting the fail2ban service, all the jails are using the same action.d file so the issue is not there.
Is there some config I am missing for manual entry or is it something else?

`fail2ban-client set jailname banip x.x.x.x`
Any help much appreciated

**Comment by sebres** [comment]
Manual ban is following the same rules as specified in jail (it was never intended to be persistent), so for example one can ban some evildoers (per script or hand) in order to avoid some attack before fail2ban is able to recognize that (filter not yet ready etc).

Although  a restart causes an unban in all jails, but there is normally a reban (restore) per default after start, so fail2ban reads from the database the active bans (`endofban > now`, where `endofban = startofban + bantime`) so if it is still active, it will be banned after restart, no matter how it was initially banned (from filter or manually).
You should see `[jail] Restore Ban <ip>` in this case.

All obsolete bans are expiring the same way (unbanned after time `startofban + bantime` reached), regardless manual or not, and restart does not change here anything.
BTW, a restart is not always necessary - you can use reload in many cases (then the tickets remains untouched, so still banned).

Because v.0.10 has a fixed bantime it is impossible at the moment to prolong the time (or somehow to change or set that as persistent bantime per ticket) at the moment.
Since v.0.11 the bantime is variable (dynamic per ticket and grows by enabled `bantime.increment` after each ban). There it were possible to specify bantime of ticket differently (I am not sure for which purposes it would be expected).

Although I don't really know why one may need that (why not directly over iptables/ipset?), but as simple "alternative" (rather as workaround) I can suggest something like this:
```bash
fail2ban-client set "$jail" banip $(
   fail2ban-regex -o 'ip' /var/log/fail2ban.log  "^[^:]*: NOTICE \[$jail\] Restore Ban <ADDR>" | tr '\n' ' '
)
```

**Comment by srulikuk** [comment]
@sebres thanks you so much, your response made me look harder and I Tested it on another few Machines and figured out the issue .
Firstly it had nothing to do with manually added bans.
my jails are configured with `action = iptables-ipset-proto6-allports-shared[name=shared,bantime=2147483]`
I did not realise that this "bantime" applies only to ipset and after a restart its no longer in fail2ban's list as it expired the default 600 to resolve i added another line in the jail `bantime = 2147483`

**Comment by sebres** [comment]
> it won't RESTORE BAN for that IP

What do you mean?
So if you execute a ban manually (e. g. configured `bantime = 15m`) and do a restart of fail2ban (e. g. not later than 5 minute hereafter), would not it restore the ban for the IP after start (at least for 10 minutes)?

If you mean my example script, then it should find the IP, because if ban occurs (no matter manually or not) fail2ban writes it to fail2ban.log (as `[<jail>] Ban <ip>`), which fail2ban-regex scans in example above.

**Comment by srulikuk** [comment]
Correct, but I MUST first repeat, this behaviour I found ONLY on ClearOS which uses Fail2Ban v0.10.5, I am not sure if ClearOS makes any changes to the code in their repo - on ubuntu 18.04 with Fail2Ban v0.10.2 I do not have this issue.

In ClearOS - currently it does NOT restore manually added if it does not have the `"fail2ban.filter [XXXXX]: INFO    [jailname] Found x.x.x.x"` entry in fail2ban.log the number of times specified in the jail.local's "maxretry=x" however this does not effect if the log is rotated (in which case it still restores).

I tested this again now by manually banning a IP and restart, it was not restored, manually banned again, added to the jail log 2 entries, restart still no restore, manually banned again added another 3 entries to the jail log (maxretry for this jail = 5) and it DID restore. (on each entry fail2ban.log added a entry FOUND)

I did not try your script as if the log is rotated it wont help, but I found when banning many IP's adding the IP to a log file is 8-9 times faster then manually banning them (I made a jail with maxretry = 1) and as soon as i add the ip to the log it gets banned, I tested this (scripted) with 108 unique IP's took 38 seconds to read IP's from DB and ban directly vs 8 seconds to read from DB and add to log file and f2b reading the log. (this is for a centralised f2b project i made with python & mysql https://github.com/srulikuk/centralised-f2b/)

**Comment by sebres** [comment]
I have no idea what could be different in ClearOS...

Anyway to restore the ban after restart, fail2ban uses sqlite-database, which should be configured (and it is per default) in fail2ban.conf (or even fail2ban.local).
If no database is configured (or no sqlite python module is installed) fail2ban is unable to restore bans after restart, so it performs scanning of the log-file or journal for the time `now - findtime` back in order to find failure entries (which are not available in case of manual ban).

**Comment by srulikuk** [comment]
I did not think to check that, my config has 
`dbfile = /var/lib/fail2ban/fail2ban.sqlite3`
`dbpurgeage = 25d` (i changed to 25d)
`dbmaxmatches = 10`
Unless the "25d" is not acceptable format (it was 1d before) the config look fine IMO

**Comment by sebres** [comment]
And does `fail2ban-python -c 'import sqlite3; print "ok"'` returns with ok?

> Is there anyway to check if its trying the DB on restore?

Well, you'd see some `Restore Ban <ip>` for all active bans.

**Comment by srulikuk** [comment]
`[root@gateway ~]# fail2ban-python -c 'import sqlite3; print "ok"'`
`ok`

Where would I see the Restore Ban? I am seeing them in the fail2ban.log but only for those that were not manually banned.

**Comment by sebres** [comment]
Hmm... it looks like manual ban is indeed not written into database (at least in 0.10).
Will check what I can do here.

**Comment by srulikuk** [comment]
> Hmm... it looks like manual ban is indeed not written into database (at least in 0.10).
As I wrote, I only found this in 0.10.5 however 0.10.2 seems to be fine.

**Comment by sebres** [comment]
Fixed in 15158e4474593aff797222fb0984658b59d5d31f for v.0.10 now, version 0.11 was not affected (the ban was added to database by observer module there).

**Comment by srulikuk** [comment]
Great, many thanks for that, I am glad I was able to get to the bottom of it and I guess help many users that don't even know they have this issue.

Will have to wait for it to trickle down to the repos of the various distros.

Once again many thanks @sebres
